### PR TITLE
Improvements to Typeclass Derivation

### PR DESCRIPTION
**Pull request by odersky**
- [x] Add a Generic instance only if the implementation of a derived
    typeclass mentions the type.

 - [x] Allow derivations for typeclasses with more or less type parameters than 1.

### CODE REVIEW

**Comment by odersky** [comment]
Only last two commits are new.

File: compiler/src/dotty/tools/dotc/typer/Deriving.scala [hunk 1]
```diff
@@ -161,33 +161,67 @@ trait Deriving { this: Typer =>
...
+
+        // Running example: typeclass: class TC[X, Y, Z], deriving class: class A[T, U]
+        // clsParamss =
+        //     T_X  T_Y  T_Z
+        //     U_X  U_Y  U_Z
+        val clsParamss: List[List[TypeSymbol]] = cls.typeParams.map { tparam =>
+          if (nparams == 0) Nil
+          else if (nparams == 1) tparam :: Nil
+          else typeClass.typeParams.map(tcparam =>
+            tparam.copy(name = s"${tparam.name}_${tcparam.name}".toTypeName)
```
[the remark this answers is not in the document]
Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/Deriving.scala]: How could that happen? Both parameter lists use disjoint names.

[the remark this answers is not in the document]
Comment (OlivierBlanvillain) [reply compiler/src/dotty/tools/dotc/typer/Deriving.scala]: - `class Foo[A_B, A]`
- `class Bar [_B_C, _C]`

yields

- A_B_B_C
- A_B_C
- A_B_C
- A_C

[the remark this answers is not in the document]
Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/Deriving.scala]: Good point. We should use a reserved separator then. Maybe `_$_`?

[the remark this answers is not in the document]
Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/Deriving.scala]: I'll do the change in the follow-up PR #5843

**Pull request merged**